### PR TITLE
IDE response to raw string literals

### PR DESCRIPTION
**Pull request by allisonchou**
Fixed the following bugs:
- Spaces added between curly brackets in raw interpolated strings (e.g. `{ x }` should be `{x}`)
- Missing completion in raw interpolated strings
- Incorrect behavior in string split handler

Added tests for:
- Completion
- Quick info
- String split command handler
- Formatting (basic case)
- Syntactic classifier (basic case)

TO-DO:
- Testing multi-braced raw interpolated strings, e.g. `$$"""{{x}}"""`. This is difficult with the current test infrastructure since `$$` is often considered the cursor position.
- Fix buggy syntactic classifier for multi-braced interpolated strings
- More formatting tests
- Handle string split command handler edge case

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Classification/ClassificationHelpers.cs [hunk 1]
```diff
@@ -145,7 +145,15 @@ private static bool IsStringToken(SyntaxToken token)
...
-                || token.IsKind(SyntaxKind.InterpolatedStringEndToken);
+                || token.IsKind(SyntaxKind.InterpolatedStringEndToken)
+                || token.IsKind(SyntaxKind.InterpolatedSingleLineRawStringStartToken)
+                || token.IsKind(SyntaxKind.InterpolatedSingleLineRawStringEndToken)
+                || token.IsKind(SyntaxKind.InterpolatedMultiLineRawStringStartToken)
+                || token.IsKind(SyntaxKind.InterpolatedMultiLineRawStringEndToken)
+                || token.IsKind(SyntaxKind.SingleLineRawStringLiteralToken)
+                || token.IsKind(SyntaxKind.SingleLineRawStringLiteralExpression)
+                || token.IsKind(SyntaxKind.MultiLineRawStringLiteralToken)
+                || token.IsKind(SyntaxKind.MultiLineRawStringLiteralExpression);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Classification/ClassificationHelpers.cs:156]: the XXXExpression cases can't be hit (tehy're not a token).

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs [hunk 2]
```diff
@@ -353,7 +353,11 @@ public static bool IsEntirelyWithinCrefSyntax(this SyntaxTree syntaxTree, int po
 
         private static bool AtEndOfIncompleteStringOrCharLiteral(SyntaxToken token, int position, char lastChar)
         {
-            if (!token.IsKind(SyntaxKind.StringLiteralToken, SyntaxKind.CharacterLiteralToken))
+            if (!token.IsKind(
+                SyntaxKind.StringLiteralToken,
+                SyntaxKind.CharacterLiteralToken,
+                SyntaxKind.SingleLineRawStringLiteralToken,
+                SyntaxKind.MultiLineRawStringLiteralToken))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs:360]: indent pls.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs [hunk 3]
```diff
@@ -364,6 +368,15 @@ private static bool AtEndOfIncompleteStringOrCharLiteral(SyntaxToken token, int
...
             }
 
+            if (token.IsKind(SyntaxKind.SingleLineRawStringLiteralToken, SyntaxKind.MultiLineRawStringLiteralToken))
+            {
+                var tokenStart = token.ToString().TakeWhile(s => s == '"');
+                startLength = tokenStart.Count();
+
+                return position == token.Span.End &&
+                    (token.Span.Length == startLength || (token.Span.Length > startLength && !token.ToString().EndsWith(string.Concat(tokenStart))));
+            }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs:378]: hrmmmmmmm  something seems fishy here.  we likely want to talk about what's going on here and if it's necessary.  i'm guessing we can answer this question without having to stringigy the token.  note: the existing code seems uber suspect to me as well :)

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs [hunk 4]
```diff
@@ -388,7 +401,10 @@ private static bool AtEndOfIncompleteStringOrCharLiteral(SyntaxToken token, int
                 token = token.GetPreviousToken(includeSkipped: true, includeDirectives: true);
             }
 
-            if (token.IsKind(SyntaxKind.StringLiteralToken))
+            if (token.IsKind(
+                SyntaxKind.StringLiteralToken,
+                SyntaxKind.SingleLineRawStringLiteralToken,
+                SyntaxKind.MultiLineRawStringLiteralToken))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs:407]:
```suggestion
            if (token.Kind() is SyntaxKind.StringLiteralToken or SyntaxKind.SingleLineRawStringLiteralToken or SyntaxKind.MultiLineRawStringLiteralToken))
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs [hunk 5]
```diff
@@ -399,7 +415,14 @@ private static bool AtEndOfIncompleteStringOrCharLiteral(SyntaxToken token, int
...
 
-            if (token.IsKind(SyntaxKind.InterpolatedStringStartToken, SyntaxKind.InterpolatedStringTextToken, SyntaxKind.InterpolatedStringEndToken))
+            if (token.IsKind(
+                SyntaxKind.InterpolatedStringStartToken,
+                SyntaxKind.InterpolatedStringTextToken,
+                SyntaxKind.InterpolatedStringEndToken,
+                SyntaxKind.InterpolatedSingleLineRawStringStartToken,
+                SyntaxKind.InterpolatedSingleLineRawStringEndToken,
+                SyntaxKind.InterpolatedMultiLineRawStringStartToken,
+                SyntaxKind.InterpolatedMultiLineRawStringEndToken))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs:425]: indent plz

File: src/VisualStudio/CSharp/Impl/LanguageService/CSharpHelpContextService.cs [hunk 6]
```diff
@@ -131,11 +131,15 @@ private string TryGetText(SyntaxToken token, SemanticModel semanticModel, Docume
             return string.Empty;
         }
 
-        private bool TryGetTextForSpecialCharacters(SyntaxToken token, out string text)
+        private static bool TryGetTextForSpecialCharacters(SyntaxToken token, out string text)
```
Comment (Youssef1313) [inline src/VisualStudio/CSharp/Impl/LanguageService/CSharpHelpContextService.cs:134]: https://github.com/dotnet/roslyn/blob/71d5b9ed55854da44a74704a566a5b5828f0942f/.editorconfig#L278-L279

Looking weird this wasn't failing the build before?

File: src/VisualStudio/CSharp/Impl/LanguageService/CSharpHelpContextService.cs [hunk 7]
```diff
@@ -131,11 +131,15 @@ private string TryGetText(SyntaxToken token, SemanticModel semanticModel, Docume
...
         {
             if (token.IsKind(SyntaxKind.InterpolatedStringStartToken) ||
                 token.IsKind(SyntaxKind.InterpolatedStringEndToken) ||
-                token.IsKind(SyntaxKind.InterpolatedStringTextToken))
+                token.IsKind(SyntaxKind.InterpolatedStringTextToken) ||
+                token.IsKind(SyntaxKind.InterpolatedSingleLineRawStringStartToken) ||
+                token.IsKind(SyntaxKind.InterpolatedSingleLineRawStringEndToken) ||
+                token.IsKind(SyntaxKind.InterpolatedMultiLineRawStringStartToken) ||
+                token.IsKind(SyntaxKind.InterpolatedMultiLineRawStringEndToken))
```
Comment (Youssef1313) [inline src/VisualStudio/CSharp/Impl/LanguageService/CSharpHelpContextService.cs:142]: I think we'd need another conditional branch for `MultiLineRawStringLiteral` and `SingleLineRawStringLiteral` here as well and co-ordinate the F1 keyword to be used with dotnet/docs team.

```csharp
if (token.IsKind(SyntaxKind.MultiLineRawStringLiteral) || token.IsKind(SyntaxKind.SingleLineRawStringLiteral))
{
    text = "RawStringLiteral_CSharpKeyword"; // or """_CSharpKeyword
    return true;
}
```

Comment (Youssef1313) [inline src/VisualStudio/CSharp/Impl/LanguageService/CSharpHelpContextService.cs:142]: Consider refactoring this long list of syntax kinds into an extension in https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/CSharpExtensions.cs

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs [hunk 8]
```diff
@@ -143,7 +143,8 @@ public static IEnumerable<SyntaxKind> GetPreprocessorKeywordKinds()
 
         public static bool IsPunctuation(SyntaxKind kind)
         {
-            return kind >= SyntaxKind.TildeToken && kind <= SyntaxKind.QuestionQuestionEqualsToken;
+            return kind >= SyntaxKind.TildeToken && kind <= SyntaxKind.QuestionQuestionEqualsToken ||
+                kind is SyntaxKind.RawInterpolationOpenToken || kind is SyntaxKind.RawInterpolationCloseToken;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:147]: note: tehse tokens kinds are gone now.  sorry!

File: src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs [hunk 9]
```diff
@@ -11336,5 +11336,125 @@ private static IntPtr NewMethod(object value)
...
+
+    private static string GetS(int y)
+    {
+        return $""""""
+            {y}
+            """""";
+    }
+}";
+
+            await TestExtractMethodAsync(code, expected);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs:11457]: one more tests with `$$"""` where there is a `{` as content as well as a brace plz.  (if you added this already, then that's great).

Note: the test should have all the braces next to each other as well.

File: src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs [hunk 10]
```diff
@@ -900,5 +900,50 @@ void M()
...
+    {
+        var v = $""""""Hello[||]there
+world
+"""""";
+    }
+}");
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.SplitStringLiteral)]
+        public void TestMissingInRawStringLiteralInterpolation_MultiBrace()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs:935]: test name seems wrong...

File: src/Workspaces/CSharp/Portable/Indentation/CSharpIndentationService.Indenter.cs [hunk 11]
```diff
@@ -73,6 +74,37 @@ private static IndentationResult GetIndentationBasedOnToken(Indenter indenter, S
...
+                var nonWhitespaceOffset = endLine.GetFirstNonWhitespaceOffset();
+                Contract.ThrowIfNull(nonWhitespaceOffset);
+                return new IndentationResult(indenter.LineToBeIndented.Start, nonWhitespaceOffset.Value);
+            }
+
+            // case 1: $"""$$
+            //          """
+            // case 2: $"""
+            //          text$$
+            //          """
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Indentation/CSharpIndentationService.Indenter.cs:94]:
```suggestion
            // case 1: $"""$$
            //         """
            // case 2: $"""
            //         text$$
            //         """
```

**Review by sharwell — CHANGES_REQUESTED** [review]
It looks like the intender is applying one too many spaces for interpolated raw strings.

File: src/Workspaces/CSharp/Portable/Indentation/CSharpIndentationService.Indenter.cs (hunk 11 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Indentation/CSharpIndentationService.Indenter.cs:94]: the indenter isn't setting where the delimiter goes here.  Instead, it's stating where the caret should go when enter is typed *above* teh delimeter.  In that case, the caret is indented the same amount as the final string.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Merging in.  Note that we still need tests around indentation and splitting of raw strings.  I will add those myself in a followup.

**Pull request merged**